<a href="https://www.kaggle.com/code/khanmdsaifullahanjar/concrete-strength-prediction-with-hypertuning?scriptVersionId=162986194" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import plotly.express as px
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_csv("/kaggle/input/concrete-strength-dataset/concrete.csv")

In [3]:
data.head(5)

Cement (component 1)(kg in a m^3 mixture)  \
0                                      540.0   
1                                      540.0   
2                                      332.5   
3                                      332.5   
4                                      198.6   

   Blast Furnace Slag (component 2)(kg in a m^3 mixture)  \
0                                                0.0       
1                                                0.0       
2                                              142.5       
3                                              142.5       
4                                              132.4       

   Fly Ash (component 3)(kg in a m^3 mixture)  \
0                                         0.0   
1                                         0.0   
2                                         0.0   
3                                         0.0   
4                                         0.0   

   Water  (component 4)(kg in a m^3 mixture)  \
0                                      162.0   
1                                      162.0   
2                                      228.0   
3                                      228.0   
4                                      192.0   

   Superplasticizer (component 5)(kg in a m^3 mixture)  \
0                                                2.5     
1                                                2.5     
2                                                0.0     
3                                                0.0     
4                                                0.0     

   Coarse Aggregate  (component 6)(kg in a m^3 mixture)  \
0                                             1040.0      
1                                             1055.0      
2                                              932.0      
3                                              932.0      
4                                              978.4      

   Fine Aggregate (component 7)(kg in a m^3 mixture)  Age (day)  strength  
0                                              676.0         28     79.99  
1                                              676.0         28     61.89  
2                                              594.0        270     40.27  
3                                              594.0        365     41.05  
4                                              825.5        360     44.30

# Data Preprocessing

## Null Values

In [4]:
df = data.copy()

In [5]:
df.isna().sum()

Cement (component 1)(kg in a m^3 mixture)                0
Blast Furnace Slag (component 2)(kg in a m^3 mixture)    0
Fly Ash (component 3)(kg in a m^3 mixture)               0
Water  (component 4)(kg in a m^3 mixture)                0
Superplasticizer (component 5)(kg in a m^3 mixture)      0
Coarse Aggregate  (component 6)(kg in a m^3 mixture)     0
Fine Aggregate (component 7)(kg in a m^3 mixture)        0
Age (day)                                                0
strength                                                 0
dtype: int64

## Duplicate Values

In [6]:
df.duplicated().sum()

25

In [7]:
df.drop_duplicates(inplace = True)

In [8]:
df.duplicated().sum()

0

## Scaling

In [9]:
from sklearn.preprocessing import MinMaxScaler

In [10]:
scaler = MinMaxScaler()

In [11]:
df.dtypes

Cement (component 1)(kg in a m^3 mixture)                float64
Blast Furnace Slag (component 2)(kg in a m^3 mixture)    float64
Fly Ash (component 3)(kg in a m^3 mixture)               float64
Water  (component 4)(kg in a m^3 mixture)                float64
Superplasticizer (component 5)(kg in a m^3 mixture)      float64
Coarse Aggregate  (component 6)(kg in a m^3 mixture)     float64
Fine Aggregate (component 7)(kg in a m^3 mixture)        float64
Age (day)                                                  int64
strength                                                 float64
dtype: object

In [12]:
for col_name in df.columns:
    if col_name != "strength":
        df[col_name] = scaler.fit_transform(df[[col_name]])

In [13]:
df.head(5)

Cement (component 1)(kg in a m^3 mixture)  \
0                                   1.000000   
1                                   1.000000   
2                                   0.526256   
3                                   0.526256   
4                                   0.220548   

   Blast Furnace Slag (component 2)(kg in a m^3 mixture)  \
0                                           0.000000       
1                                           0.000000       
2                                           0.396494       
3                                           0.396494       
4                                           0.368392       

   Fly Ash (component 3)(kg in a m^3 mixture)  \
0                                         0.0   
1                                         0.0   
2                                         0.0   
3                                         0.0   
4                                         0.0   

   Water  (component 4)(kg in a m^3 mixture)  \
0                                   0.321086   
1                                   0.321086   
2                                   0.848243   
3                                   0.848243   
4                                   0.560703   

   Superplasticizer (component 5)(kg in a m^3 mixture)  \
0                                            0.07764     
1                                            0.07764     
2                                            0.00000     
3                                            0.00000     
4                                            0.00000     

   Coarse Aggregate  (component 6)(kg in a m^3 mixture)  \
0                                           0.694767      
1                                           0.738372      
2                                           0.380814      
3                                           0.380814      
4                                           0.515698      

   Fine Aggregate (component 7)(kg in a m^3 mixture)  Age (day)  strength  
0                                           0.205720   0.074176     79.99  
1                                           0.205720   0.074176     61.89  
2                                           0.000000   0.739011     40.27  
3                                           0.000000   1.000000     41.05  
4                                           0.580783   0.986264     44.30

# Data Spliting

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
X = df.drop("strength",axis=1)

In [16]:
y = df["strength"]

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Model Training

In [18]:
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
import xgboost as xgb
from xgboost import XGBRegressor

In [19]:
lr = LinearRegression()
rfr = RandomForestRegressor()
xgb = XGBRegressor()

In [20]:
## Fit

In [21]:
lr.fit(X_train,y_train)

LinearRegression()

In [22]:
rfr.fit(X_train,y_train)

RandomForestRegressor()

In [23]:
xgb.fit(X_train,y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=None, n_jobs=None,
             num_parallel_tree=None, random_state=None, ...)

In [24]:
## predict

In [25]:
predicted_lr_train = lr.predict(X_train)

In [26]:
predicted_lr_test = lr.predict(X_test)

In [27]:
predicted_rfr_train = rfr.predict(X_train)

In [28]:
predicted_rfr_test = rfr.predict(X_test)

In [29]:
predicted_xgb_train = xgb.predict(X_train)

In [30]:
predicted_xgb_test = xgb.predict(X_test)

# Scoring

In [31]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [32]:
# Mean Absolute Error (MAE)
mae = mean_absolute_error(y_test, predicted_lr_test)
print(f'Mean Absolute Error (MAE): {mae}')

# Mean Squared Error (MSE)
mse = mean_squared_error(y_test, predicted_lr_test)
print(f'Mean Squared Error (MSE): {mse}')

# Root Mean Squared Error (RMSE)
rmse = np.sqrt(mse)
print(f'Root Mean Squared Error (RMSE): {rmse}')

# R-squared (R2) score
r2 = r2_score(y_test, predicted_lr_test)
print(f'R-squared (R2) score: {r2}')

Mean Absolute Error (MAE): 8.984068581174695
Mean Squared Error (MSE): 126.20929107403164
Root Mean Squared Error (RMSE): 11.234290857639019
R-squared (R2) score: 0.560915111117331


In [33]:
# Mean Absolute Error (MAE)
mae = mean_absolute_error(y_test, predicted_rfr_test)
print(f'Mean Absolute Error (MAE): {mae}')

# Mean Squared Error (MSE)
mse = mean_squared_error(y_test, predicted_rfr_test)
print(f'Mean Squared Error (MSE): {mse}')

# Root Mean Squared Error (RMSE)
rmse = np.sqrt(mse)
print(f'Root Mean Squared Error (RMSE): {rmse}')

# R-squared (R2) score
r2 = r2_score(y_test, predicted_rfr_test)
print(f'R-squared (R2) score: {r2}')

Mean Absolute Error (MAE): 3.82834006859035
Mean Squared Error (MSE): 29.92607427308122
Root Mean Squared Error (RMSE): 5.4704729478429215
R-squared (R2) score: 0.895886531925905


In [34]:
# Mean Absolute Error (MAE)
mae = mean_absolute_error(y_test, predicted_xgb_test)
print(f'Mean Absolute Error (MAE): {mae}')

# Mean Squared Error (MSE)
mse = mean_squared_error(y_test, predicted_xgb_test)
print(f'Mean Squared Error (MSE): {mse}')

# Root Mean Squared Error (RMSE)
rmse = np.sqrt(mse)
print(f'Root Mean Squared Error (RMSE): {rmse}')

# R-squared (R2) score
r2 = r2_score(y_test, predicted_xgb_test)
print(f'R-squared (R2) score: {r2}')

Mean Absolute Error (MAE): 3.1672845013725834
Mean Squared Error (MSE): 23.230438446360438
Root Mean Squared Error (RMSE): 4.819796515036754
R-squared (R2) score: 0.9191807956679459


# Hyperparameter Tuning

In [35]:
from sklearn.model_selection import GridSearchCV

In [36]:
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2],
}


In [37]:
grid_search = GridSearchCV(estimator=xgb, param_grid=param_grid, cv=5, scoring='r2')
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=XGBRegressor(base_score=None, booster=None,
                                    callbacks=None, colsample_bylevel=None,
                                    colsample_bynode=None,
                                    colsample_bytree=None, device=None,
                                    early_stopping_rounds=None,
                                    enable_categorical=False, eval_metric=None,
                                    feature_types=None, gamma=None,
                                    grow_policy=None, importance_type=None,
                                    interaction_constraints=None,
                                    learning_rate=None, m...
                                    max_cat_threshold=None,
                                    max_cat_to_onehot=None, max_delta_step=None,
                                    max_depth=None, max_leaves=None,
                                    min_child_weight=None, missing=nan,
                                    monotone_constraints=None,
                                    multi_strategy=None, n_estimators=None,
                                    n_jobs=None, num_parallel_tree=None,
                                    random_state=None, ...),
             param_grid={'learning_rate': [0.01, 0.1, 0.2],
                         'max_depth': [3, 5, 7],
                         'n_estimators': [50, 100, 200]},
             scoring='r2')

In [38]:
best_params = grid_search.best_params_

In [39]:
best_xgb_regressor = XGBRegressor(**best_params)

In [40]:
best_xgb_regressor.fit(X_train,y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.2, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=5, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=200, n_jobs=None,
             num_parallel_tree=None, random_state=None, ...)

In [41]:
predicted = best_xgb_regressor.predict(X_test)

In [42]:
# Mean Absolute Error (MAE)
mae = mean_absolute_error(y_test, predicted)
print(f'Mean Absolute Error (MAE): {mae}')

# Mean Squared Error (MSE)
mse = mean_squared_error(y_test, predicted)
print(f'Mean Squared Error (MSE): {mse}')

# Root Mean Squared Error (RMSE)
rmse = np.sqrt(mse)
print(f'Root Mean Squared Error (RMSE): {rmse}')

# R-squared (R2) score
r2 = r2_score(y_test, predicted)
print(f'R-squared (R2) score: {r2}')

Mean Absolute Error (MAE): 3.0564892466968256
Mean Squared Error (MSE): 22.040202464981448
Root Mean Squared Error (RMSE): 4.694699400918172
R-squared (R2) score: 0.9233216527251445


# Evaluation

1. **Linear Regression:**
   - Moderate performance with relatively high MAE, MSE, and RMSE.
   - R2 score suggests that the model explains only 56% of the variance in the concrete strength.

2. **Random Forest Regressor:**
   - Improved performance compared to Linear Regression.
   - Lower MAE, MSE, and RMSE, indicating better predictive accuracy.
   - High R2 score (0.90) indicates a good fit to the data.

3. **Gradient Boosting Regressor (XGBoost):**
   - Further improvement over Random Forest with lower MAE, MSE, and RMSE.
   - High R2 score (0.92) suggests excellent explanatory power.
   - A robust model for concrete strength prediction.

4. **XGBoost After Hyperparameter Tuning:**
   - Slight improvement in MAE, MSE, and RMSE.
   - R2 score increased to 0.92, indicating enhanced model performance.
   - Suggests that hyperparameter tuning refined the model.


# Conclusion:

- The Gradient Boosting Regressor (XGBoost) demonstrates the best performance for concrete strength prediction, with or without hyperparameter tuning.
- The hyperparameter tuning of XGBoost leads to marginal improvement, indicating that the initial model configuration was already well-tuned.
- Random Forest also performs well but falls slightly behind XGBoost in predictive accuracy.